code/modules/mining/machine_vending.dm

```diff
@@ -126,62 +106,89 @@
+		data["has_id"] = FALSE
+
+/obj/machinery/mineral/equipment_vendor/tgui_static_data(mob/user)
+	var/list/static_data[0]
+	. = static_data
+
+	// Available items - in static data because we don't wanna compute this list every time! It hardly changes.
+	static_data["items"] = list()
+	for(var/cat in prize_list)
+		var/list/cat_items = list()
+		for(var/prize_name in prize_list[cat])
+			var/datum/data/mining_equipment/prize = prize_list[cat][prize_name]
+			cat_items[prize_name] = list("name" = prize_name, "price" = prize.cost)
+		static_data["items"][cat] = cat_items
+
+/obj/machinery/mineral/equipment_vendor/vv_edit_var(var_name, var_value)
+	// Gotta update the static data in case an admin VV's the items for some reason..!
+	if(var_name == "items")
+		dirty_items = TRUE
+
+/obj/machinery/mineral/equipment_vendor/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
+	// Update static data if need be
+	if(dirty_items)
+		if(!ui)
+			ui = SStgui.get_open_ui(user, src, ui_key)
+		ui?.initial_static_data = tgui_static_data(user)
+		dirty_items = FALSE
+
+	// Open the window
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "MiningVendor", name, 400, 450)
+		ui.open()
+		ui.set_autoupdate(FALSE)
+
+/obj/machinery/mineral/equipment_vendor/tgui_act(action, params)
 	if(..())
-		return 1
+		return
 
-	if(href_list["choice"])
-		if(istype(inserted_id))
-			if(href_list["choice"] == "eject")
-				inserted_id.loc = loc
-				inserted_id.verb_pickup()
-				inserted_id = null
-		else if(href_list["choice"] == "insert")
-			var/obj/item/card/id/I = usr.get_active_hand()
-			if(istype(I))
-				if(!usr.drop_item())
-					return
-				I.loc = src
-				inserted_id = I
-			else
-				to_chat(usr, "<span class='danger'>No valid ID.</span>")
-
-	if(href_list["purchase"])
-		if(istype(inserted_id))
-			var/datum/data/mining_equipment/prize = locate(href_list["purchase"])
-			if(!prize || !(prize in prize_list) || prize.cost > inserted_id.mining_points)
+	. = TRUE
+	switch(action)
+		if("logoff")
+			if(!inserted_id)
+				return
+			usr.put_in_hands(inserted_id)
+			inserted_id = null
+		if("purchase")
+			if(!inserted_id)
+				return
+			var/category = params["cat"] // meow
+			var/name = params["name"]
+			if(!(category in prize_list) || !(name in prize_list[category])) // Not trying something that's not in the list, are you?
+				return
+			var/datum/data/mining_equipment/prize = prize_list[category][name]
+			if(prize.cost > inserted_id.mining_points) // shouldn't be able to access this since the button is greyed out, but..
+				to_chat(usr, "<span class='danger'>You have insufficient points.</span>")
 				return
 
 			inserted_id.mining_points -= prize.cost
-			new prize.equipment_path(src.loc)
-	updateUsrDialog()
+			new prize.equipment_path(loc)
+		else
+			return FALSE
+	add_fingerprint()
```

Comment:
Might wanna add this to the start of the `tgui_act` because currently it wont be hit

Reply:
It will, since the actual cases don't return. Only the default case returns, which makes sense because you don't want to add fingerprints if no action is performed (due to it being non-existent)

Reply:
Oh I see, github was squishing my diff view :v